tool/compute-training-coverage.js

```diff
@@ -0,0 +1,140 @@
+// limitations under the License.
+//
+// Author: Silei Xu <silei@cs.stanford.edu>
+"use strict";
+
+const fs = require('fs');
+const ThingTalk = require('thingtalk');
+const Tp = require('thingpedia');
+
+const Utils = require('../lib/utils/misc-utils');
+const { requoteProgram } = require('../lib/dataset-tools/requoting');
+const { readAllLines } = require('./lib/argutils');
+const { DatasetParser } = require('../lib/dataset-tools/parsers');
+
+function waitEnd(stream) {
+    return new Promise((resolve, reject) => {
+        stream.on('end', resolve);
+        stream.on('error', reject);
+    });
+}
+
+async function normalize(preprocessed, target_code, schemas) {
+    const entities = Utils.makeDummyEntities(preprocessed);
+    const sequence = target_code.split(' ');
+    const parsed = ThingTalk.NNSyntax.fromNN(sequence, entities);
+    await parsed.typecheck(schemas);
+    const normalized = ThingTalk.NNSyntax.toNN(parsed, preprocessed, {}, {
+        allocateEntities: true,
+        typeAnnotations: false
+    });
+    return normalized.join(' ');
+}
+
+
+module.exports = {
+    initArgparse(subparsers) {
+        const parser = subparsers.addParser('compute-training-coverage', {
+            addHelp: true,
+            description: "Given a evaluation set, compute the percent of programs appeared in training set"
+        });
+        parser.add_argument('--thingpedia', {
+            required: true,
+            help: 'Path to ThingTalk file containing class definitions.'
+        });
+        parser.add_argument('-l', '--locale', {
+            required: false,
+            default: 'en-US',
+            help: `BGP 47 locale tag of the language to evaluate (defaults to 'en-US', English)`
+        });
+        parser.addArgument('--training-set', {
+            required: true,
+            type: fs.createReadStream,
+            help: 'Path to the file containing the training data (in TSV format: id, utterance, thingtalk)'
+        });
+        parser.addArgument('--evaluation-set', {
+            required: true,
+            type: fs.createReadStream,
+            help: `Path to the file containing the evaluation data (in TSV format: id, utterance, thingtalk)`
+        });
+    },
+
+    async execute(args) {
+        const tpClient = new Tp.FileClient(args);
+        const schemas = new ThingTalk.SchemaRetriever(tpClient, null, true);
+
+        let trainingSize = 0;
+        const trainingPrograms = new Map(); // counter of unique programs in training set
+        args.training_set.setEncoding('utf8');
+        const training = await readAllLines([args.training_set])
+            .pipe(new DatasetParser({ preserveId: true }));
+        training.on('data', async (line) => {
+            const normalized = await normalize(line.preprocessed, line.target_code, schemas);
+            const requoted = Array.from(requoteProgram(normalized)).join(' ');
+            trainingPrograms[requoted] = (trainingPrograms[requoted] || 0) + 1;
```

Comment:
This is not how you use a Map ;)

Reply:
All fixed in the last commit. (I will just do a squash merge)